sophus/so3.hpp

```diff
@@ -261,32 +271,31 @@ class SO3Base {
     /// Representation through Encapsulation of Manifolds"
     /// Information Fusion, 2011
 
-    if (squared_n <
-        Constants<Scalar>::epsilon() * Constants<Scalar>::epsilon()) {
+    if (squared_n < Constants<Scalar>::epsilon() * Constants<Scalar>::epsilon()) {
       // If quaternion is normalized and n=0, then w should be 1;
       // w=0 should never happen here!
       SOPHUS_ENSURE(abs(w) >= Constants<Scalar>::epsilon(),
                     "Quaternion (%) should be normalized!",
                     unit_quaternion().coeffs().transpose());
-      Scalar squared_w = w * w;
-      two_atan_nbyw_by_n =
-          Scalar(2) / w - Scalar(2.0 / 3.0) * (squared_n) / (w * squared_w);
-      J.theta = Scalar(2) * squared_n / w;
+      theta_by_n =
+          Scalar(2) / w - Scalar(2.0/3.0) * (squared_n) / (w * squared_w);
+      J.theta = theta_by_n * n;
     } else {
-      Scalar n = sqrt(squared_n);
-      if (abs(w) < Constants<Scalar>::epsilon()) {
-        if (w > Scalar(0)) {
-          two_atan_nbyw_by_n = Constants<Scalar>::pi() / n;
-        } else {
-          two_atan_nbyw_by_n = -Constants<Scalar>::pi() / n;
-        }
+      // Use computationally stable (around theta=pi, w=0) version of atan2(y, x)
+      // with (w = x) >= 0, (n = y) > 0
+      /*
+      if (w < Constants<Scalar>::epsilon()) {
+        // it fails with zero gradients in ceres optimization
+        J.theta = Constants<Scalar>::pi();
       } else {
-        two_atan_nbyw_by_n = Scalar(2) * atan(n / w) / n;
+        J.theta = Scalar(4) * atan(n / (w + sqrt(squared_w + squared_n)));
       }
-      J.theta = two_atan_nbyw_by_n * n;
-    }
+      */
+      J.theta = Scalar(4) * atan(n / (w + sqrt(squared_w + squared_n)));
 
-    J.tangent = two_atan_nbyw_by_n * unit_quaternion().vec();
+      theta_by_n = J.theta / n;
+    }
+    J.tangent = theta_by_n * sign * unit_quaternion().vec();
```

Comment:
It is better to use `std::copysign`:
```cpp
copysign(theta_by_n, w) * unit_quaternion().vec();
```

Reply:
Thanks for the comment.
I am afraid that the `copysign` is not reloaded for `ceres::Jet` type. 

It gives the following error during compilation: 
```
error: no matching function for call to ‘copysign(
    Sophus::SO3Base<Sophus::SO3<ceres::Jet<double, 3> > >::Scalar&, 
    Sophus::SO3Base<Sophus::SO3<ceres::Jet<double, 3> > >::Scalar&)’
J.tangent = copysign(theta_by_n, w) * unit_quaternion().vec();
```

Reply:
I've submitted a Jet implementation of `copysign` [here](https://ceres-solver-review.googlesource.com/c/ceres-solver/+/19220).

Reply:
`copysign` support for ceres `Jet` has now been[ merged](https://github.com/ceres-solver/ceres-solver/commit/a3a4b6d77f376815ac7475a7d2e5b6df41797300).